src/combinatorics/hall/basic.lean

```diff
@@ -0,0 +1,230 @@
+      have subi : ({i} : finset ι) ⊆ {i,i'} := by simp,
+      have subi' : ({i'} : finset ι) ⊆ {i,i'} := by simp,
+      have le : ∀ {s t : finset ι}, s ⊆ t → s ≤ t := λ _ _ h, h,
+      rw [←hu (category_theory.hom_of_le (le subi)).op,
+          ←hu (category_theory.hom_of_le (le subi')).op],
+      let uii' := u (opposite.op ({i,i'} : finset ι)),
+      exact λ h, subtype.mk_eq_mk.mp (uii'.property.1 h), },
+    { intro i,
+      apply (u (opposite.op ({i} : finset ι))).property.2, }, },
+  { rintro ⟨f, hf₁, hf₂⟩ s,
+    rw ←finset.card_image_of_injective s hf₁,
+    apply finset.card_le_of_subset,
+    intro _,
+    rw [finset.mem_image, finset.mem_bUnion],
+    rintros ⟨x, hx, rfl⟩,
+    exact ⟨x, hx, hf₂ x⟩, },
+end
+
+/-- Given a relation such that the image of every singleton set is finite, then the image of every
+finite set is finite. -/
+instance {α β : Type*} [decidable_eq β]
+  (r : α → β → Prop) [∀ (a : α), fintype (rel.image r {a})]
+  (A : finset α) : fintype (rel.image r A) :=
+begin
+  have h : rel.image r A = (A.bUnion (λ a, (rel.image r {a}).to_finset) : set β),
+  { ext, simp [rel.image], },
+  rw [h],
+  apply finset_coe.fintype,
+end
+
+/--
+This is a version of **Hall's Marriage Theorem** in terms of a relation
+between types `α` and `β` such that `α` is finite and the image of
+each `x : α` is finite (it suffices for `β` to be finite; see
+`fintype.all_card_le_filter_rel_iff_exists_injective`).  There is
+a transversal of the relation (an injective function `α → β` whose graph is
+a subrelation of the relation) iff every subset of
+`k` terms of `α` is related to at least `k` terms of `β`.
+
+If `[fintype β]`, then `[∀ (a : α), fintype (rel.image r {a})]` is automatically implied.
+-/
+theorem fintype.all_card_le_rel_image_card_iff_exists_injective
+  {α β : Type*} [decidable_eq β]
+  (r : α → β → Prop) [∀ (a : α), fintype (rel.image r {a})] :
+  (∀ (A : finset α), A.card ≤ fintype.card (rel.image r A)) ↔
+    (∃ (f : α → β), function.injective f ∧ ∀ x, r x (f x)) :=
+begin
+  let r' := λ a, (rel.image r {a}).to_finset,
+  have h : ∀ (A : finset α), fintype.card (rel.image r A) = (A.bUnion r').card,
+  { intro A,
+    rw ←set.to_finset_card,
+    apply congr_arg,
+    ext b,
+    simp [rel.image], },
+  have h' : ∀ (f : α → β) x, r x (f x) ↔ f x ∈ r' x,
+  { simp [rel.image], },
+  simp only [h, h'],
+  apply finset.all_card_le_bUnion_card_iff_exists_injective,
+end
+
+/--
+This is a version of **Hall's Marriage Theorem** in terms of a relation to a finite type.
+There is a transversal of the relation (an injective function `α → β` whose graph is a subrelation
+of the relation) iff every subset of `k` terms of `α` is related to at least `k` terms of `β`.
+
+It is like `fintype.all_card_le_rel_image_card_iff_exists_injective` but uses `finset.filter`
+rather than `rel.image`.
+-/
+theorem fintype.all_card_le_filter_rel_iff_exists_injective
+  {α β : Type*} [fintype β]
+  (r : α → β → Prop) [∀ a, decidable_pred (r a)] :
```

Comment:
Sad that this can't be `decidable_rel r` 😢 
Maybe put a TODO so that we think about it when `decidable_rel` gets pulled off core and generalized?

Reply:
I'm adding this note:
```
/- TODO: decidable_pred makes Yael sad. When an appropriate decidable_rel-like exists, fix it. -/
```
I don't think it matters so much, since they're all abbreviations for `decidable` in the end so it doesn't affect users of this theorem.

I suggest that instead of generalizing `decidable_rel`, which has the sort of nice constraint that the two arguments have the same type, that there be a suite of abbreviations `decidable₁`, `decidable₂`, and `decidable₃` for 1, 2, and 3 arguments.  Maybe they could even be for dependent functions for maximum generality.  These can be created now without needing to alter lean core.  (Putting this note here since it seems like something you might be excited about doing.)